File: ghc-4335c07/nofib/real/fem/VBlldecomp.hs
-- Glasgow Haskell 0.403 : BLAS ( Basic Linear Algebra System )
-- **********************************************************************
-- *                                                                    *
-- * FILE NAME : vblldecomp.hs		DATE : 5-3-1991                 *
-- *                                                                    *
-- * CONTENTS : LL decomposition method for variable bandwidth matrix   *
-- * 									*
-- **********************************************************************

module VBlldecomp(vblldecomp, vbllsolution, vbllsolution') where

import Basics

import Vector

import VBmatrix

vblldecomp :: Vbm Float -> Vbm Float 

	-- L[i,i] = sqrt(A[i,i] - ***)
	--          where
	--	    *** = SUM [ L[i,k]*L[i,k] | k <- [1..i-1] ]
	-- L[i,j] = (A[i,j] - ***) / L[j,j]
	--	    where
	--	    *** = SUM [ L[i,k]*L[j,k] | k <- [1..j-1] ]

vbllsolution :: Vbm Float -> Vec Float -> Vec Float
	-- Solve equation system, the variable bandwidth matrix is 
	-- undecomposed, ie it is the original matrix.

vbllsolution' :: Vbm Float -> Vec Float -> Vec Float
        -- Solve equation system, the variable bandwidth matrix is decomposed.

vblldecomp mA = m
        where 
        m = makevbmat (boundvbmat mA) (diagadrvbm mA) f 
        f (i,j) =
		if (i == j) then 
			sqrt ( vbmatsub mA (i,i)  -
			       sum ( map ( \ k -> vbmatsub m (i,k) *
                                           vbmatsub m (i,k) )
			  	        [(fstclvbmat mA i) .. (i-1)] ) )

		else 
		    (	vbmatsub mA (i,j) -
			sum ( map ( \ k -> (vbmatsub m (i,k) *
                                            vbmatsub m (j,k) ) )
				  [ (k0 i j) .. (j - 1)] )
                    ) / ( vbmatsub m (j,j) )
	k0 i j = if ( (fstclvbmat mA i) >= (fstclvbmat mA j) ) then
	   	     (fstclvbmat mA i)
	         else 
		     (fstclvbmat mA j)

vbllsolution a b =
	fst (backwarding ( forwarding (b, vblldecomp a) ) )

vbllsolution' a' b =
	fst (backwarding ( forwarding (b, a') ) )

forwarding ( b,  mVB ) =
	(b',mVB)
	where
	b' = makevec n f
	n = boundvec b
	f i = ( vecsub b i -
		sum [(vbmatsub mVB (i,j)) * (vecsub b' j) | j<-[l i..i-1] ] 
              ) / (vbmatsub mVB (i,i))
	l i = fstclvbmat mVB i

backwarding ( b, mVB ) =
	(b',mVB)
	where
	n = boundvec b
	b' = makevec n f
	f i = ( vecsub b i -
		sum [( vbmatsub mVB (j,i) ) * ( vecsub b' j )
                     | j <- ( validj i [i+1..n] ) ]
              ) / (vbmatsub mVB (i,i))
	validj i (j:js) = if ( i >= fstclvbmat mVB j ) then 
				j : (validj i js)
                          else 
				validj i js
	validj i []      = []

File: ghc-4335c07/nofib/real/fem/inC/vblldecomp.c
/*variable bandwidth matrix ll decomposition , and system solution */

#include <math.h>

vblldecomp(n,diag,vbm)

int n,diag[];
float  vbm[];

{
  int i,j,k,adr,j0,k0;
  float  t,x;

  for (i=1; i<=n; i=i+1) {
    
    j0 = i - diag[i] + diag[i-1] + 1; /* column of first non zero elemnt */

    for (j=j0; j<i; j=j+1) {
      
      t = 0.0; 
      k0 = j - diag[j] + diag[j-1] + 1; /* fst.non.zero.col of jth row */
      if (j0>k0) k0 = j0;
      for (k=k0; k<=j-1; k=k+1) {
        x = vbm[diag[i]+k-i] * vbm[diag[j]+k-j];
        t = t + x;
      }
      adr = diag[i]+j-i;
      vbm[adr] = (vbm[adr]-t)/vbm[diag[j]];
  
    }

    t = 0.0; 
    for (k=j0; k<=(i-1); k=k+1) {
      x = vbm[diag[i]+k-i];
      t = t + x*x;
    }
    adr = diag[i];
    vbm[adr] = sqrt ( vbm[adr] - t );

  }
  

}
  

vbllsolution(n,diag,mvb,b)

int n, diag[];
float  mvb[], b[];

{
  int i,j,j0;
  float  t;

  for (i=1; i<=n; i=i+1) {
    t = 0.0;
    j0 = i - diag[i] + diag[i-1] + 1;
    for (j=j0; j<i; j=j+1) {
      t = t + b[j] * mvb[diag[i]+j-i];
    }
    b[i] = (b[i]-t) / mvb[diag[i]];
  }


  for (i=n; i>=1; i=i-1) {
    b[i] = b[i] / mvb[diag[i]];
    j0 = i - diag[i] + diag[i-1] + 1;
    for (j=j0; j<i; j=j+1) {
      b[j] = b[j] - b[i]*mvb[diag[i]+j-i];
    }
  }


}
